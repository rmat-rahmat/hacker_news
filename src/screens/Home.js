import React from "react";
import { View, Text, StyleSheet, ActivityIndicator, Modal, Pressable } from "react-native";
import List from "./list";
import Details from "./details";
import getTopNews from "../api/getTopNews";
import getNewsDetails from "../api/getNewsDetails";
import { chunk } from "../api/utilities";

const Home = () => {

    const [data, setData] = React.useState(null);
    const [activeData, setActiveData] = React.useState(null);
    const [loadGroup, setLoadGroup] = React.useState(1);
    const [detailData, setDetailData] = React.useState(null);
    const [loading, setLoading] = React.useState(true);
    const [processing, setProcessing] = React.useState(false);

    React.useEffect(() => {
        loadData();
    }, []);

    const loadData = () => {
        setProcessing(true);

        // Get news from API
        getTopNews().then(x => {

            // Split large array into multiple array of 10 item
            const d = chunk(x, 10);
            setData(d);

            // Set initial 20 item to load
            const initialList = [...d[0], ...d[1]];

            fetchData(initialList).then(n => {
                setActiveData(n);
                setProcessing(false);
                if (loading) setLoading(false);
            }).catch(e=>console.log(e));
        }).catch(e => console.log(e));
    }

    // Fetch item detail from API
    const fetchData = (x) => {
        return new Promise(async (resolve, reject) => {
            let data = [];

            // Using async loop to prevent server timeout
            for (const key of x) {
                const contents = await getNewsDetails(key);
                data.push(contents);
            }
            resolve(data);
        });


    }

    // add more item to data list
    const loadMore = () => {

        // prevent data fetch if the process still run
        if (loadGroup < data.length && !processing) {
            const n = loadGroup + 1;

            setProcessing(true);
            fetchData(data[n]).then(ft => {
                setActiveData([...activeData, ...ft]);
                setProcessing(false);
            })
            setLoadGroup(n);
        }
    }

    // set parent item to view comment
    const setIndex = x => {
        const filterData = activeData.find(({ id }) => id === x);
        setDetailData(filterData);
    }

    return (
        <View style={styles.container}>
            <View style={styles.header}>
                {detailData && <Pressable onPress={() => setDetailData(null)}><Text style={styles.headerText}>{("< Back ")}</Text></Pressable>}
                <Text style={styles.headerText}>Hacker News</Text>
            </View>
            <View style={styles.content} >
                {activeData ? detailData ?
                    <Details data={detailData} /> :
                    <List data={activeData} setIndex={setIndex} reloadData={loadData} loadMore={loadMore} processing={processing} /> :
                    <Text>Data not found</Text>
                }
                <Modal
                    animationType="slide"
                    transparent={true}
                    visible={loading}
                >
                    <View style={styles.centeredView}>
                        <ActivityIndicator size="large" />
                        <Text style={styles.title}>Loading...</Text>
                    </View>
                </Modal>
            </View>
        </View>
    );
};

export default Home;

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: "column",
        height: '100%',
    },
    header: {
        backgroundColor: "orange",
        flex: 0.08,
        alignItems: 'center',
        flexDirection: "row",
    },
    headerText: {
        fontWeight: 'bold',
        fontSize: 20,
        color: 'black',
        paddingHorizontal: 10
    },
    content: {
        backgroundColor: "white",
        flex: 0.92
    },
    item: {
        backgroundColor: '#f9c2ff',
        padding: 20,
        marginVertical: 8,
        marginHorizontal: 16,
    },
    title: {
        color: 'black',
        fontSize: 32,
    },
    centeredView: {
        flex: 1,
        justifyContent: "center",
        alignItems: "center",
        marginTop: 22
    }
});