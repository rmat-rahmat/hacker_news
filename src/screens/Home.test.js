/**
 * @format
 */

import 'react-native';
import React from 'react';
import Home from './Home';
import getTopNews from "../api/getTopNews";

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve([1, 10, 100, 1000, 10000, 100000]),
  })
);
beforeEach(() => {
  fetch.mockClear();
});

it('Fetch mock data', () => {
  getTopNews().then(x=>{
    expect(x).toHaveLength(6);;
  })
  expect(fetch).toHaveBeenCalledTimes(1);
});

it('renders correctly', () => {
  const tree = renderer.create(
    <Home />
    ).toJSON();
  expect(tree).toMatchSnapshot();
});
