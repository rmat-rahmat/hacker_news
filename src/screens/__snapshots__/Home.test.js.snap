// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "orange",
        "flex": 0.08,
        "flexDirection": "row",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 20,
          "fontWeight": "bold",
          "paddingHorizontal": 10,
        }
      }
    >
      Hacker News
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "flex": 0.92,
      }
    }
  >
    <Text>
      Data not found
    </Text>
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      transparent={true}
      visible={true}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "marginTop": 22,
          }
        }
      >
        <ActivityIndicator
          size="large"
        />
        <Text
          style={
            Object {
              "color": "black",
              "fontSize": 32,
            }
          }
        >
          Loading...
        </Text>
      </View>
    </Modal>
  </View>
</View>
`;
